### ISSUE TITLE
Router top-K selection has no guard for best==-1: a single NaN logit causes out-of-bounds reads and writes at index -1

### ISSUE THREAD

**Issue by mxkev1n** [JustVugg/colibri#558]
### Summary

All three top-K selection loops in `moe()` initialise `best=-1` and never check it before
using it as an index. If every candidate score is NaN, no comparison ever succeeds, `best`
stays `-1`, and the code performs an out-of-bounds read **and three out-of-bounds heap
writes plus a stack buffer underflow**.

### Affected

- `c/colibri.c:2881-2885` (plain top-K path)
- `c/colibri.c:2802-2806` and `c/colibri.c:2813-2817` (`CACHE_ROUTE` ranking loops — same pattern)
- Commit `81f08a0`, v1.1.1. CPU router path, any build.

### Detail

```c
for(int kk=0;kk<Ksel;kk++){ int best=-1; float bv=-1e30f;
    for(int e=0;e<E;e++){ int tk=0; for(int j=0;j<kk;j++) if(idx[j]==e){tk=1;break;}
        if(!tk && choice[e]>bv){bv=choice[e];best=e;} }
    idx[kk]=best; w[kk]=logit[best];          // <-- best may be -1
}
```

`choice[e] = sigmoidf(logit[e]) + l->router_bias[e]`. `sigmoidf(NaN)` is NaN, and
`NaN > bv` is false for every `e`, so `best` remains `-1`. Then:

1. `w[kk]=logit[best]` reads `logits_all[s*E - 1]` — heap OOB read when `s==0`.
2. `idx[kk]=-1` flows into the accounting loop a few lines below:
   ```c
   m->eusage[layer][idx[kk]]++;
   m->eheat[layer][idx[kk]]++;
   m->elast[layer][idx[kk]]=++m->eaccess_clock;
   ```
   — **three OOB writes at index -1** on separate `calloc`'d arrays.
3. It then reaches FASE B: `if(!seen[e]){ seen[e]=1; ... }` writes `seen[-1]`, a **stack
   buffer underflow** on the VLA `unsigned char seen[E]`.
4. Finally `uniq[nu++] = -1` is handed to `expert_load()`, which formats it into a tensor
   name and fails to find it.

### Why this is reachable, not theoretical

The repo already treats non-finite logits as a real, observed condition. From
`c/tests/test_logit_nan.c`:

> A single NaN or +Inf in the logits (**a bad streamed expert tile, or an fp overflow in the
> matmul at a low-RAM eviction boundary**) used to make softmax produce an all-NaN g_pbuf...

That regression hardened the **sampling** side (`argmax_v` skips NaN, `dist_build` collapses
to one-hot). The **routing** side was not hardened, and it is upstream of sampling — a NaN
that reaches the router corrupts memory before it can ever reach `argmax_v`. The stated
threat model (streamed expert tiles, low-RAM eviction boundaries) is exactly this engine's
normal operating regime.

Note the author is clearly aware of the pattern: the `EXPERT_BUDGET` selection loop a few
hundred lines below **does** have the guard (`if(best<0) break;`).

### Suggested fix

Guard all three loops, and range-check the pre-routed shortcut too:

```diff
-        idx[kk]=best; w[kk]=logit[best];
+        if(best<0){                       /* NaN logits or exhausted candidates */
+            static int warned; if(!warned){ warned=1;
+                fprintf(stderr,"[router] non-finite logits at layer %d — degrading\n",layer); }
+            best = kk<E ? kk : 0;         /* deterministic fallback, never a bad index */
+        }
+        idx[kk]=best; w[kk]=logit[best];
```

Also worth adding once after the `g_pre_idx` memcpy: assert `0 <= idxs[i] < E` and
`keff[s] <= K`, so a bad GPU-side routing result fails loudly instead of corrupting the same
arrays.

### Verification

Run under ASan/UBSan with a deliberately NaN-poisoned hidden state (or an int3/E8 tensor
carrying an inf pattern) and confirm the OOB write on `m->eusage[layer][-1]` and on `seen[-1]`.

**Comment by JustVugg** [comment]
Fixed and merged to `dev` (#563). Confirmed the bug against the source: all three top-K scans seed `best=-1`, and a single NaN logit (`sigmoidf(NaN)=NaN`, `NaN>bv` false for every expert) leaves `best==-1` — then `logit[-1]` OOB read plus OOB writes to `eusage/eheat/elast[layer][-1]` and a `seen[-1]` VLA underflow. Real, and reachable in the streamed-tile / low-RAM eviction regime.

The fix guards all three loops (healthy path is byte-for-byte unchanged; only `best<0` is intercepted, warns once, returns an in-range fallback) and also sanitizes the pre-routed GPU shortcut. It ships `tests/test_router_nan.c`, now wired into the suite — I ran it, it passes. Thank you for the report and the fix.
